### DISCUSSION TITLE
Adding custom Http Handlers

### DISCUSSION THREAD

**Discussion by vic0824** [ArcadeData/arcadedb#600] [Q&A] [answered]
I'm working on a project that consists of several components, the main one contains an http server that processes requests and interacts with a database.
The database I currently use is Hsqldb, but I'd like to replace it with ArcadeDB to handle larger data volumes.
Since ArcadeDB already uses Undertow as an http server, I think it would be a waste of resources if I started another http server in parallel with the ArcadeDB http server.
I believe Undertow will have no problems managing the few extra requests that my server needs to handle (order of magnitude is 10 requests per minute).
I was wondering if there is a way of adding some custom http handlers to the public class HttpServer created by ArcadeDBServer without modifying the ArcadeDB code.
The ArcadeDBServer class exposes a getHttpServer() method that returns the HttpServer instance, but HttpServer does not expose any method to add custom handlers or to add another Http Listener.
I have seen that the ServerPlugin interface contains a registerAPI method (that accepts a PathHandler as an argument), which the HttpServer.startService method invokes for every registered plugin. I have therefore thought about re-writing each of my handlers as implementations of ServerPlugin. However, I have seen that all the ArcadeDB classes that implement ServerPlugin contain an empty implementation of registerAPI, so maybe it hasn't been designed with this particular use case in mind.
Any suggestions?

**Comment by lvca** [comment, marked as answer]
@vic0824 I think that using ArcadeDB's Undertow is an excellent idea! You're right, the easiest way is to create a server plugin (look at `MongoDBProtocolPlugin` plugin implementation) and then implement the `registerAPI` method. Example for the HTTP POST API `/myapi/test`:
```
@Override
public void registerAPI(HttpServer httpServer, final PathHandler routes) {
routes.addPrefixPath("/myapi",//
  routes.post("/test/{par1}/{par2}", new MyTestAPI(this));
}
```

At startup, ArcadeDB Server will initiate your plugin and register your API. To start the server with your plugin, register the full class in 
`arcadedb.server.plugins` setting:

Example:

`java ... -Darcadedb.server.plugins=MyPlugin:com.yourpackage.MyPlugin ...`

**Reply by vic0824** [reply]
Great! thanks.

**Comment by vic0824** [comment]
Unfortunately I cannot get it to work.
The MongoDBProtocolPlugin class contains an empty registerAPI method (and so do all other *Plugin classes).
The example in the doc uses "routes.post" wich I assume is a typo, as PathHandler does not have a "post" method, and it should be Handlers.routing().post).
The example also contains "new MyTestAPI(this)", but "this" would be an instance of MyTest, i.e. a ServerPlugin, while the MyTestApi constructor in the example requires an HttpServer instance.
I have therefore looked at the HttpServer class as an example.

This is how the registerAPI method looks like in my custom plugin:
@Override
    public void registerAPI(HttpServer httpServer, PathHandler routes) {
        routes
                .addPrefixPath("/orchestrator", Handlers.routing().get("/retrieveProcessors", new RetrieveProcessorsHandler(httpServer)));
    }

When I start the server (embedded in my application), the log says that the plugin was started:
2022-12-25 17:15:04.736 INFO  [ArcadeDBServer] <ArcadeDB_0> - Orchestrator plugin started
but all my http requests are rejected with a  404 status.
Another thing I don't understand: I use the server in development mode, so if it receives a URL that does not match any of the configured routes, it should use the fallback handler which serves the Studio webpage.
For example, when I send the following http request with a browser: 
http://192.168.163.128:8080/orchestrator/retrieveProcessors
I get: "Failed to load resource: the server responded with a status of 404 (Not Found)"
The IP and port number are correct, because if I use "http://192.168.163.128:8080/" as an URL, I get the Studio login page.
There seems to be no undertow log file, so I don't know how to obtain more information about why Undertow rejects the request.
I attach the plugin class and the handler class I have used as an example.
[OrchestratorPlugin.zip](https://github.com/ArcadeData/arcadedb/files/10300404/OrchestratorPlugin.zip)
[RetrieveProcessorsHandler.zip](https://github.com/ArcadeData/arcadedb/files/10300405/RetrieveProcessorsHandler.zip)

**Reply by lvca** [reply]
There was a typo in the docs, your example seems right to me, it's like all the other API we register for HTTP and Studio. I'm going to check out your sources tomorrow.

**Reply by lvca** [reply]
One difference I can see is that you redefine the prefix per each API, while we're using this form:

```java
routes.addPrefixPath("/api/v1",//
        basicRoutes//
            .post("/begin/{database}", new PostBeginHandler(this))//
            .post("/close/{database}", new PostCloseDatabaseHandler(this))//
            .post("/command/{database}", new PostCommandHandler(this))//
            .post("/commit/{database}", new PostCommitHandler(this))//
            .post("/create/{database}", new PostCreateDatabaseHandler(this))//
```

**Reply by vic0824** [reply]
That was it! Thanks @lvca.
It is a really cool feature. As an added bonus, your GetDynamicContentHandler allows me to serve my static content without writing a single line of code.

**Reply by lvca** [reply]
Glad it worked! I hope more people use ArcadeDB server as their server. Maybe we need a nice blog post about it. Any volunteer? :-)

**Reply by lvca** [reply]
In the meanwhile I explained a little better that section of the docs (should be online in <10 mins)

**Reply by vic0824** [reply]
Sure, I can write a post in "Show and tell". Did you have something else in mind?

**Reply by lvca** [reply]
Sounds good to me

**Reply by vic0824** [reply]
done: https://github.com/ArcadeData/arcadedb/discussions/711

**Reply by lvca** [reply]
Thanks so much!
